### ISSUE TITLE
unexpected line in Next():

### ISSUE THREAD

**Issue by bussiere** [ipfs/go-ipfs#1688]
I'am currently manipulating pictures and gallery of pictures :  
Seems kind of backslash plague with \\n\
```
   time="16:33:05:000" level=error msg="multipart: unexpected line in Next(): \"A\\xa1טY\\xfa\\x12\\xe4\\xdbĬ \\x8bOr\\x80u\\x86\\xc2\\x03X\\fj\\x97\\x86\\x9dd\\xb18\\x14\\x85\\x84#\\xe1\\x16\\xe6d\\x00\\xb6\\a2j\\xf8})\\xb1ļ\\x10P\\b{e[2T\\xc361\\xaf\\x12\\xb6\\xf1\\xb2>\\xd3r\\xc4S5\\xc4/\\xa4W)O<\\xc3\\x1d\\xca\\xc0\\x03\\xa4\\xddW\\x830\\xcdv\\xfcõq<\\x92\\xdf\\x19\\x89ឝ`\\xa6Q\\xab\\xe5nZ\\xb5q\\xb4\\x81\\x980\\x11\\x93R\\xe7QR+l\\xf9\\x98\\xcd_\\x04\\xa4@;`\\n\"" module="commands/http


ipfs add -r . > result.txt
time="16:33:05:000" level=error msg="multipart: unexpected line in Next(): \"A\\xa1טY\\xfa\\x12\\xe4\\xdbĬ\\x8bOr\\x80u\\x86\\xc2\\x03X\\fj\\x97\\x86\\x9dd\\xb18\\x14\\x85\\x84#\\xe1\\x16\\xe6d\\x00\\xb6\\a2j\\xf8})\\xb1ļ\\x10P\\b{e[2T\\xc361\\xaf\\x12\\xb6\\xf1\\xb2>\\xd3r\\xc4S5\\xc4/\\xa4W)O<\\xc3\\x1d\\xca\\xc0\\x03\\xa4\\xddW\\x830\\xcdv\\xfcõq<\\x92\\xdf\\x19\\x89ឝ`\\xa6Q\\xab\\xe5nZ\\xb5q\\xb4\\x81\\x980\\x11\\x93R\\xe7QR+l\\xf9\\x98\\xcd_\\x04\\xa4@;`\\n\"" module="commands/http" 
bussiere@kusanagi:/media/bussiere/ext_kus111/_build$ leafpad result.txt 
bussiere@kusanagi:/media/bussiere/ext_kus111/_build$ ipfs add -r . > result.txt
time="17:10:11:000" level=error msg="unexpected EOF" module="commands/http" 
bussiere@kusanagi:/media/bussiere/ext_kus111/_build$ 
```

**Comment by whyrusleeping** [comment]
@bussiere what version of ipfs (git sha if you can), OS and archictecture?

**Comment by whyrusleeping** [comment]
Also, is there anything weird about the gallery? such as uf8 file names or strange file types

**Comment by bussiere** [comment]
Ubuntu 14.04 go 1.5 amd64
~/Lib/Go/src/github.com/jbenet/go-ipfs$ git rev-parse HEAD
a0e5fc8294c2551f3a8c599f2a78ed5ba4167534

And yes in the gallerie it may have some french character as ç or japanese one.

Thanks

**Comment by bussiere** [comment]
That's strange because if i try to add it folder by folder it seems to works :

**Comment by whyrusleeping** [comment]
@bussiere do you mind zipping up a set of files that can repro the issue?

**Comment by bussiere** [comment]
yep the zip is here : QmfYQRPWworETuBudtayhkmCAMxmz8LV3YXm3KXwytyrfM

and i've isolated the error ::


bussiere@kusanagi:/media/bussiere/ext_kus111/101APPLE$ ipfs add -r .
added QmV6BSMWFohrBXVHBEP21asUr8LFJZyz157BtPyNLmwzzs 101APPLE/IMG_1750.JPG/5003.JPG
added QmVtC61fyphzwbBDQ5zGoZVddg4H6D2hRvdewHim2j9ou4 101APPLE/IMG_1750.JPG/index.html
added Qmdp2JAxw7s6mAAsJYdoa8DTjqtkBU2jgWQSRwthGkKx7y 101APPLE/IMG_1750.JPG/thumbnails/5003.JPG
added QmbaQJLPh2jc99BTN6qhdcJT9uMZUdU76jYpDgC8TUDLtp 101APPLE/IMG_1750.JPG/thumbnails
added QmdtitSecheXBnR4KWY6WLxcbVYY6oYeGBC7gJoN6QWXmu 101APPLE/IMG_1750.JPG
added QmX68ToAb8EBssnoT1NcqQwU4oYyxiFViF4F9jrneqAmwE 101APPLE/IMG_1751.JPG/5003.JPG
added QmQVCbVXRXutwQmC9NBubAvnf7QnqyKtFzE6cW1iMkn8Gf 101APPLE/IMG_1751.JPG/index.html
added QmaxjEAnbkwtyW3WHc9HdxSQvBBmUTx5NKLN5oMAJXJAzY 101APPLE/IMG_1751.JPG/thumbnails/5003.JPG
added QmbMcBeVikQPiroUxGFbpknLFthgScrEuBkfktNfACHonu 101APPLE/IMG_1751.JPG/thumbnails
added QmRdQZZEjhe1ZdTb13udjNm6CBrveh8H8Cg2ctRetBKVBu 101APPLE/IMG_1751.JPG
added QmZfLYsLgmSQx7JiscQCHLwoMS68uxTeNMbHK3CGZHUeJW 101APPLE/IMG_1752.JPG/5003.JPG
added QmaKScnQyqguLF1pDmo1fNowMnhAMuoEKt2H4U9jrmiqWu 101APPLE/IMG_1752.JPG/index.html
ERRO[19:28:53:000] multipart: unexpected line in Next(): "\x97\xec(\x0eQ\x80\xeb\u007f16)\x8a\xf1\n"  module=commands/http
added QmbFMke1KXqnYyBBWxB74N4c5SBnJMVAiMNRcGu6x1AwQH 101APPLE/IMG_1752.JPG/thumbnails/5003.JPG

**Comment by jbenet** [comment]
multipart blowing up?

**Comment by ion1** [comment]
Just like #1722, @eminence reported that this only happens with the binary from gobuilder.me but not with one he built.

**Comment by jbenet** [comment]
Which version is the gobuilder one? 0.3.7 or 0.3.8-dev?



—

**Comment by eminence** [comment]
The gobuilder version I downloaded came from the URL below and was downloaded approximately at 5:00 PM EST today.   THe .zip archive has a timestamp of "Sept 16 02:41", and it identifies itself as `ipfs version 0.3.8-dev`.

This version string is identical to the string produced by the working version of IPFS that doesn't reproduce this issue

https://gobuilder.me/get/github.com/ipfs/go-ipfs/cmd/ipfs/ipfs_master_linux-amd64.zip

**Comment by jbenet** [comment]
SIgh, ok. i worried we'd run into a case like this soon enough. we need to start putting git commit short hashes into the version string (`0.3.8-dev-a1b2c3`) or something.

@whyrusleeping thoughts here?

**Comment by sonatagreen** [comment]
I'm getting the same error. I pulled a fresh version a few minutes ago with `go get -u github.com/ipfs/go-ipfs/cmd/ipfs`, and `ipfs version` returns `ipfs version 0.3.8-dev`.

I've attached a file that causes the error. I experimented a bit with how much I could 'simplify' the file while still getting the error, in order to try to narrow down what's happening.

The following seems not to affect whether the error occurs:

* replacing a normal character with another normal character
* reordering characters

But the following does:

* adding or deleting a normal or unusual character
* replacing a normal character with an unusual character, or vice versa

Replacing an unusual character with another unusual character sometimes affects it, and sometimes doesn't. I haven't worked out the pattern.

I've identified the following characters as 'unusual':

* U+00A0 NO-BREAK SPACE
* U+00A9 COPYRIGHT SIGN
* U+2013 EN DASH
* U+2261 IDENTICAL TO

I assume there are many others, which I haven't bothered to look for.

**Comment by davidar** [comment]
Could this be related to #1582 ?

**Comment by sonatagreen** [comment]
Possible, but I don't see how bad UTF-8 handling would explain the fact that the crash goes away when I change the leading characters from `©©©©––` to `©©©–––`.

**Comment by davidar** [comment]
To be honest I'm more surprised that ipfs doesn't crash on binary files more often given the broken json codec.

**Comment by whyrusleeping** [comment]
@davidar broken json codec? i may be missing some context here

**Comment by davidar** [comment]
@whyrusleeping the issue I just linked, with the json encoding corrupting binary data

**Comment by jbenet** [comment]
how common is this for you @sonatagreen ?

**Comment by sonatagreen** [comment]
It's only happened to me once so far.

**Comment by sonatagreen** [comment]
Ran into it again today, so twice now.

One-liner to reproduce:

    gunzip /ipfs/QmXhHWLZcT4BbVrihUZK6bm9Detty9vAztmiMR2M9kP58g/crashfile.txt.gz -c > crashfile.txt ; ipfs add crashfile.txt

**Comment by eminence** [comment]
i ran into this today.  i didn't have the file that triggered this error

**Comment by mhe** [comment]
I ran into the same issue when experimenting with IPFS. I started to dig into it a bit and the interesting thing is that it appears to depend on *file size*. I can reproduce this behaviour for any file of size n*4032-1 (where n is an integer > 0).

    dd if=/dev/zero of=testfile bs=1 count=8063
    ipfs add testfile
    ipfs cat QmPftE3vHBA9cVS1DgnFXLhv4reSvLzEKbCWqCYe94zafU > testfile_cat
    diff testfile testfile_cat
    hexdump testfile
    hexdump testfile_cat

The hex dump of testfile:

    0000000 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
    * 
    0001f70 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00   
    0001f7f

The hex dump of testfile_cat:

    0000000 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
    *
    0001f70 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 0d
    0001f80 0a 2d 2d                                       
    0001f83

Considering the "boundary-line" of the mime-multipart thing appears to be of length 62, together with two characters \r and \n this adds up to 64. Add this to the magic number 4032 (see above) and you have 4096. This happens to be the same as the peek buffer size in go's multipart package (https://golang.org/src/mime/multipart/multipart.go). I haven't had the time to pinpoint the exact cause, but I suspect the cause of this bug is hiding in the complexity of (the peek buffer part of) this multipart package. Also considering the diff between the original and the retrieved file: it is 4 bytes longer, exactly '\r\n--'. Perhaps the boundary-line is not properly detected in these cases? Anyway, perhaps this is another data-point that can help in resolving this.

**Comment by jbenet** [comment]
>  I can reproduce this behaviour

@mhe great! could you contribute a sharness test case so we can make it pass?

**Comment by mhe** [comment]
@jbenet Sure, sounds like an interesting exercise :) I've attached a sharness test to this comment (I thought an PR would be a bit overkill).

[t0250-multipart-file4031.sh.txt](https://github.com/ipfs/go-ipfs/files/41040/t0250-multipart-file4031.sh.txt)

**Comment by jbenet** [comment]
Ok thanks @mhe -- we should make a PR with a fix that includes that test case.

**Comment by travisperson** [comment]
I assume this is fixed as of `1.5.2` golang/go/issues/12662 ?

**Comment by whyrusleeping** [comment]
@travisperson ah yes! thanks, i can close this now.
